### PR TITLE
In-place transform for sparse data with variances

### LINKED ISSUE #216: Support arithmetic operations for sparse data
- (Binary) Operations between sparse and dense data.
- After #250, we need to support propagation of uncertainties for sparse data in the `transform` helpers.

### LINKED ISSUE #264: Complete and improve set of transform functions for Variable
So far we have mainly `transform_in_place` with two arguments.

- We need also non-in-place variants (see #272), and at least an idea how to provide variants with more arguments.
- We need to support propagation of uncertainties also for the sparse operator used internally.
- The single-argument version does not appear to support non-contiguous data in the current implementation.

Once this is complete we can more easily do #213 and #216. Furthermore, we would then be ready for extremely easy addition of more operations.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
This is a major iteration of the `tranform` (or more specifically `transfom_in_place`) functions. Depends on #271.

Major changes:
- Support propagation of uncertainties for sparse data.
- Cleaner syntax for functors and `transform_in_place` parameter order (in-place variant is not relying on return value any more). This also avoid expensive copies in the case of sparse data.
- Specifying things like `sparse_container<double>` when calling `transform_in_place` is not necessary any more. Code for handling both sparse and dense data is not generated automatically.
- Some code documentation.

Fixes #264 and fixes #216 (mostly, at least).

There will definitely be a couple of follow-ups to this PR:
- Testing is very incomplete and needs major refactor and additions.
- Some operations for sparse data with variances not supported yet, such as appending (using `insert` and `begin` and `end` of a proxy iterator (essentially what `ValuesAndVariances` is).

### CODE REVIEW

File: core/transform.h [hunk 1]
```diff
@@ -57,108 +113,264 @@ constexpr auto operator*(const ValueAndVariance<T1> a, const T2 b) noexcept {
...
+  }
+  template <class T1, class T2>
+  constexpr void operator()(sparse_container<T1> &a,
+                            const sparse_container<T2> &b) const {
+    expect::sizeMatches(a, b);
+    transform_in_place_impl(op, a, b);
+  }
+  template <class T1, class T2>
+  constexpr void operator()(ValuesAndVariances<T1> a,
+                            const sparse_container<T2> b) const {
```
Comment (nvaytet) [inline core/transform.h]:
```suggestion
                            const sparse_container<T2> &b) const {
```

File: core/transform.h [hunk 2]
```diff
@@ -57,108 +113,264 @@ constexpr auto operator*(const ValueAndVariance<T1> a, const T2 b) noexcept {
...
       if (b.hasVariances()) {
         auto b_var = b.variances();
-        for (scipp::index i = 0; i < a_val.size(); ++i) {
-          const ValueAndVariance a_{a_val[i], a_var[i]};
-          const ValueAndVariance b_{b_val[i], b_var[i]};
-          const auto out = op(a_, b_);
-          c_val[i] = out.value;
-          c_var[i] = out.variance;
-        }
+        transform_in_place_with_variance_impl(op,
```
Comment (nvaytet) [inline core/transform.h]: :+1: nice

File: core/transform.h [hunk 3]
```diff
@@ -13,11 +32,48 @@ namespace scipp::core {
...
 };
 
+template <class T>
+constexpr auto operator-(const ValueAndVariance<T> a) noexcept {
+  return ValueAndVariance{-a.value, a.variance};
+}
+
+template <class T> constexpr auto sqrt(const ValueAndVariance<T> a) noexcept {
+  using std::sqrt;
+  return ValueAndVariance{sqrt(a.value), 0.25 * (a.variance / a.value)};
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply core/transform.h]: You mean INF? How?

File: core/transform.h (hunk 1 above)
Comment (SimonHeybrock) [reply core/transform.h]: :+1, but this overload is removed in a follow-up branch, not changing here.

File: core/test/transform_test.cpp [hunk 4]
```diff
@@ -34,90 +41,338 @@ TEST(Variable, apply_unary) {
...
+      {sparse_container<double>(2), sparse_container<double>(2)});
+  Variable val_var = a;
+  Variable val = makeVariable<double>(
+      dims, {sparse_container<double>(2), sparse_container<double>(2)});
+  static constexpr auto op = [](auto &a, const auto b) { a *= b; };
+};
+
+TEST_F(TransformInPlaceTest_sparse_binary_values_variances_size_fail,
+       baseline) {
+  ASSERT_NO_THROW(transform_in_place<pair_self_t<double>>(a, val_var, op));
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply core/test/transform_test.cpp]: Well, as the name says it is the "baseline". The other tests in the fixture fail. Without having a baseline we would not know that the other tests fail due to the expected reason.

**Pull request merged**